x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,163 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer func() {
+		_ = l.Close()
+	}()
+	addr := address.Address(l.Addr().String())
+
+	testCases := []struct {
+		desc              string
+		dialer            func(Dialer) Dialer
+		handshaker        func(Handshaker) Handshaker
+		expectErr         bool
+		expectPoolCleared bool
+	}{
+		{
+			desc:              "no errors",
+			dialer:            nil,
+			handshaker:        nil,
+			expectErr:         false,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "dialer context timeout",
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(_ context.Context, network, address string) (net.Conn, error) {
+					ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(-1*time.Hour))
+					defer cancel()
+					return d.DialContext(ctx, network, address)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "dialer context cancelled",
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(_ context.Context, network, address string) (net.Conn, error) {
+					ctx, cancel := context.WithCancel(context.Background())
+					cancel()
+					return d.DialContext(ctx, network, address)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "GetHandshakeInformation context timeout",
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(_ context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(-1*time.Hour))
+						defer cancel()
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "GetHandshakeInformation context canceled",
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(_ context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						ctx, cancel := context.WithCancel(context.Background())
+						cancel()
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Start a goroutine that pulls events from the events channel and inserts them into the
+			// events slice. Use a sync.WaitGroup to allow the test code to block until the events
+			// channel loop exits, guaranteeing that all events were copied from the channel.
+			events := make([]*event.PoolEvent, 0)
+			eventsCh := make(chan *event.PoolEvent)
+			var eventsWg sync.WaitGroup
+			eventsWg.Add(1)
+			go func() {
```

Comment:
Could you have this go routine just set pool cleared instead of storing the events?

Reply:
That would technically work for checking if there was/wasn't a pool clear event. However, the test also relies on a few more behaviors provided by using the `channel`:
- Expect that there is at least 1 event emitted, otherwise fail
- Expect that no events are emitted after the events channel is closed, otherwise panic